### ISSUE TITLE
Problem with GraphQL Introspection query...

### ISSUE THREAD

**Issue by Panoplos** [SocketCluster/sc-formatter#1]
Performing the following query:

```graphql

  query IntrospectionQuery {
    __schema {
      queryType { name }
      mutationType { name }
      subscriptionType { name }
      types {
        ...FullType
      }
      directives {
        name
        description
        locations
        args {
          ...InputValue
        }
      }
    }
  }

  fragment FullType on __Type {
    kind
    name
    description
    fields(includeDeprecated: true) {
      name
      description
      args {
        ...InputValue
      }
      type {
        ...TypeRef
      }
      isDeprecated
      deprecationReason
    }
    inputFields {
      ...InputValue
    }
    interfaces {
      ...TypeRef
    }
    enumValues(includeDeprecated: true) {
      name
      description
      isDeprecated
      deprecationReason
    }
    possibleTypes {
      ...TypeRef
    }
  }

  fragment InputValue on __InputValue {
    name
    description
    type { ...TypeRef }
    defaultValue
  }

  fragment TypeRef on __Type {
    kind
    name
    ofType {
      kind
      name
      ofType {
        kind
        name
        ofType {
          kind
          name
          ofType {
            kind
            name
            ofType {
              kind
              name
              ofType {
                kind
                name
                ofType {
                  kind
                  name
                }
              }
            }
          }
        }
      }
    }
  }
```
I am getting this error in `sc-formatter`:

```shell
project: sockets   operation name: IntrospectionQuery
project: sockets   variables
undefined
1506911535944 - Origin: Worker(PID 4117)
[Warning] TypeError: Cannot convert object to primitive value
at Array.binaryToBase64Replacer(/home/panoplos/Documents/Development/Node.js/project/node_modules/sc-formatter/index.js:40:29)
at JSON.stringify(<anonymous>)
    at Object.module.exports.encode (/home/panoplos/Documents/Development/Node.js/project/node_modules/sc-formatter/index.js:88:15)
    at SCSocket.encode (/home/panoplos/Documents/Development/Node.js/project/node_modules/socketcluster-server/scsocket.js:258:28)
    at SCSocket.sendObjectSingle (/home/panoplos/Documents/Development/Node.js/project/node_modules/socketcluster-server/scsocket.js:289:16)
    at SCSocket.sendObject (/home/panoplos/Documents/Development/Node.js/project/node_modules/socketcluster-server/scsocket.js:302:10)
    at Response._respond (/home/panoplos/Documents/Development/Node.js/project/node_modules/socketcluster-server/response.js:15:17)
    at Response.end (/home/panoplos/Documents/Development/Node.js/project/node_modules/socketcluster-server/response.js:27:10)
    at Response.callback (/home/panoplos/Documents/Development/Node.js/project/node_modules/socketcluster-server/response.js:51:10)
    at SCSocket.<anonymous> (/home/panoplos/Documents/Development/Node.js/project/src/server/sockets/scGraphQLHandler.js:22:3)
    at <anonymous>
      at process._tickCallback (internal/process/next_tick.js:188:7)
```

Has anyone else had this issue? I am using the following as the GraphQL handler socket:

```es6
import { execute } from 'graphql';
import rootSchema from 'server/graphql/rootSchema'

export default async (exchange, socket, debug) => {
  const schema = await rootSchema()
  return async (data, callback) => {
    const { query, variables } = data;
    debug.log(`Handling a graphql ${query.definitions[0].operation} from socket [${socket.id}]`)
    debug.log(`operation name: ${!!query.definitions[0].name ? query.definitions[0].name.value : undefined}`)
    debug.log(`variables`, variables)
    const authToken = socket.authToken;
    const context = {
      authToken,
      exchange,
      socket
    };
    const result = await execute(schema, query, {}, context, variables);
    if (result.errors) {
      debug.error('DEBUG GraphQL Error:', result.errors);
    }
    callback(null, result);
  }
}
```
Where `rootSchema` is returning a schema created with `makeExecutableSchema` from `graphql-tools`.
Note that all other queries succeed.

Additionally, `introspectionQueury` from `graphql/utilities` also fails, however, when I introspect the endpoint via an HTTP interface, as follows, there are no issues.

```es6
...
  if (process.env.NODE_ENV !== 'production') {
    debug.log(`We also put up nice convenient GraphQL endpoints on /graphql and /graphiql ...`)
    const { graphqlKoa, graphiqlKoa } = require('apollo-server-koa')
    const getRootSchema = require('server/graphql/rootSchema').default
    getRootSchema().then((rootSchema) => {
      router.all('/graphql', graphqlKoa( ctx => ({
        schema: rootSchema,
        context: { authToken: ctx.get('Authorization').split('Bearer ')[1] }
      })))
      router.all('/graphiql', graphiqlKoa({
        endpointURL: '/graphql'
      }))
    })
  }
...
```

**Comment by Panoplos** [comment]
Is this repo dead?

**Comment by jondubois** [comment]
@Panoplos Thanks for raising this issue in the main SC repo. Sorry for the delay; for some reason, I wasn't watching this repo on GitHub. Can you make a simple script that I can setup quickly in Node.js with SocketCluster to reproduce this issue? (with as few dependencies as possible).

**Comment by jondubois** [comment]
@Panoplos The query result you're sending back is in a string right?

**Comment by jondubois** [comment]
I think this might be happening when it tries to send the result of the query back to the client.
possibly on the line `callback(null, result);` in your code. Can you check what is the `result` from `const result = await execute(schema, query, {}, context, variables);`?

**Comment by Panoplos** [comment]
Note that I have been using this for a while with no issues on other queries. It is just this introspection query that is failing.

The result is in javascript object (JSON) format (not text) and is pretty large -- and contains proprietary information. How should we go about this? Setting up a test repo that scaffolds both client and server SocketCluster is not trivial, so best mechanism would be to get you the schema result data directly to test.

**Comment by jondubois** [comment]
@Panoplos what is your sc-formatter version? `npm ls sc-formatter`
It looks like it might be related to a `Buffer` (binary) which is inside your response object somewhere.

Also Node.js version?

**Comment by jondubois** [comment]
@Panoplos is the problem in sc-formatter the line `if (value && value.type == 'Buffer' && value.data instanceof Array) {`? (that's line 40 according to the error)

If so, does changing it to:

```
if (value && value.type == 'Buffer' && Array.isArray(value.data)) {
```

fix the problem (just guessing)?

**Comment by Panoplos** [comment]
```
└─┬ socketcluster@8.0.2
  ├─┬ sc-broker-cluster@4.3.0
  │ └─┬ sc-broker@4.1.0
  │   └─┬ ncom@1.0.0
  │     └── sc-formatter@3.0.0  deduped
  └─┬ socketcluster-server@8.0.0
    └── sc-formatter@3.0.0  deduped
```

Node: v8.8.0 (but this existed through a couple versions)

I'll test that in a sec...

**Comment by Panoplos** [comment]
That change didn't help.

Here is the data it is choking on:
```
args:Array(0) []
deprecationReason:null
description:"@HACK Need to provide a no-op here or schema generator balks..."
isDeprecated:false
name:"dummy"
type:Object {kind: "SCALAR", name: "String", ofType: null}
  kind:"SCALAR"
  name:"String"
  ofType:null
```

**Comment by Panoplos** [comment]
OK, the following works:

```es6
var binaryToBase64Replacer = function (key, value) {
  if (global.ArrayBuffer && value instanceof global.ArrayBuffer) {
    return {
      base64: true,
      data: arrayBufferToBase64(value)
    };
  } else if (global.Buffer) {
    if (value instanceof global.Buffer){
      return {
        base64: true,
        data: value.toString('base64')
      };
    }
    // Some versions of Node.js convert Buffers to Objects before they are passed to
    // the replacer function - Because of this, we need to rehydrate Buffers
    // before we can convert them to base64 strings.
    if (value && value.type === 'Buffer' && Array.isArray(value.data)) {
      var rehydratedBuffer;
      if (global.Buffer.from) {
        rehydratedBuffer = global.Buffer.from(value.data);
      } else {
        rehydratedBuffer = new global.Buffer(value.data);
      }
      return {
        base64: true,
        data: rehydratedBuffer.toString('base64')
      };
    }
  }
  return value;
};
```

**Comment by jondubois** [comment]
@Panoplos Note that line 25 is not the same as line 40.

`value instanceof global.ArrayBuffer` vs `value.data instanceof Array` (one is ArrayBuffer, the other is Array)

**Comment by Panoplos** [comment]
Correct, 25 left as is.... 40 needs the strict comparison. Sorry, it is too late here...brain quickly turning to mush. (Updated the comment.)

**Comment by Panoplos** [comment]
Also, `value.data instanceof Array` works, so technically just need that one character change... :laughing:

**Comment by jondubois** [comment]
@Panoplos Are you sure that the `===` is the issue? I don't see why it would make a difference.

**Comment by jondubois** [comment]
Ah yes. I can reproduce something similar:

```
var s = '{"x": 5, "toString": "foo"}';
var d = JSON.parse(s);
if (d == 'abc') {
}
```

But using `===` instead fixes the issue because it doesn't try to downcast the object into a string.

```
var s = '{"x": 5, "toString": "foo"}';
var d = JSON.parse(s);
if (d === 'abc') {
}
```

In any case it means that there is something strange about the data which you're returning (in the case above, we're overriding the `toString` method. Maybe the object prototype was overriden in a strange way or it was constructed abnormally. E.g. `Object.create(null) == ''` also causes the issue.

I'll update sc-formatter to use `===` instead anyway.

**Comment by jondubois** [comment]
@Panoplos I've published the fix in sc-formatter v3.0.1. Thanks for reporting this issue.
